{% extends 'base.html' %}

{% block title %}Sources{% endblock %}

{% block content %}
<h1>Sources</h1>
    <div id="d_sources_div">
        <h2 style="font-weight: bold">Disclaimer</h2>
        <p>
            The use of PhyloRNA is free of charge. The authors have taken all necessary steps to ensure that the database and website are implemented correctly and perform the function as described in the documentation. However, neither the authors nor the University of Camerino guarantee the accuracy, suitability for a particular purpose, or future availability of PhyloRNA.
        </p>
        <hr>
        <h2 style="font-weight: bold">Condition of use</h2>
        <p>
            PhyloRNA consists of {{ num_docs }} RNA Secondary Structures from 8 different sources. For each RNA sequence, we provide the corresponding bpseq file, fasta file, dbn file and ct file. Moreover, the csv file of each RNA secondary structure is available. All of the files are available to download after doing a search.
        </p>
        <ul>
            <li><p>RCSB Protein Data Bank: <a href="http://www.rcsb.org/pdb/">http://www.rcsb.org/pdb/</a></p></li>
            <li><p>Gutell Lab CRW Site: <a href="http://www.rna.ccbb.utexas.edu/">http://www.rna.ccbb.utexas.edu/</a></p></li>
            <li><p>tmRNA Database: <a href="http://rnp.uthct.edu/rnp/tmRDB/tmRDB.html">http://rnp.uthct.edu/rnp/tmRDB/tmRDB.html</a></p></li>
            <li><p>Sprinzl tRNA Database: <a href="http://www.uni-bayreuth.de/departments/biochemie/sprinzl/trna/">http://www.uni-bayreuth.de/departments/biochemie/sprinzl/trna/</a></p></li>
            <li><p>RNase P Database: <a href="http://jwbrown.mbio.ncsu.edu/RNaseP/home.html">http://jwbrown.mbio.ncsu.edu/RNaseP/home.html</a></p></li>
            <li><p>SRP Database: <a href="http://rnp.uthct.edu/rnp/SRPDB/SRPDB.html">http://rnp.uthct.edu/rnp/SRPDB/SRPDB.html</a></p></li>
            <li><p>Nucleic Acid Database: <a href="http://ndbserver.rutgers.edu/">http://ndbserver.rutgers.edu/</a></p></li>
            <li><p>Rfam Data: <a href="http://www.sanger.ac.uk/Software/Rfam/index.shtml">http://www.sanger.ac.uk/Software/Rfam/index.shtml</a></p></li>
        </ul>
        <br>
        <p>
            Additionally, we request that you cite the appropriate publication listed below as well as the original sources of the PhyloRNA data that you used, which are stated above.
        </p>
        <ul>
            <li><p>[STRAND] Andronescu M, Bereg V, Hoos HH, and Condon A: RNA STRAND: The RNA Secondary Structure and Statistical Analysis Database. BMC Bioinformatics. 2008;9(1):340.</p></li>
            <li><p>[PDB] Westbrook J, Feng Z, Chen L, Yang H, Berman H: The Protein Data Bank and structural genomics. Nucleic Acids Res 2003, 31:489-491.</p></li>
            <li><p>[CRW] Cannone J, Subramanian S, Schnare M, Collett J, D'Souza L, Du Y, Feng B, Lin N, Madabusi L, Muller K, Pande N, Shang Z, Yu N, Gutell R: The comparative RNA web (CRW) site: an online database of comparative sequence and structure information for ribosomal, intron, and other RNAs. BMC Bioinformatics 2002, 3:2. [Correction: BMC Bioinformatics 3:15.]</p></li>
            <li><p>[tmRDB,SRPDB] Andersen ES, Rosenblad MA, Larsen N, Westergaard JC, Burks J, Wower IK, Wower J, Gorodkin J, Samuelsson T, Zwieb C: The tmRDB and SRPDB resources. Nucleic Acids Res 2006, 34(Database issue):163-168.</p></li>
            <li><p>[SprinzlDB] Sprinzl M, Vassilenko K: Compilation of tRNA sequences and sequences of tRNA genes. Nucleic Acids Res 2005, 33(Database issue):139-140.</p></li>
            <li><p>[RNasePDB] Brown J: The Ribonuclease P Database. Nucleic Acids Res 1999, 27:314-314.</p></li>
            <li><p>[Rfam] Griffiths-Jones S, Moxon S, Marshall M, Khanna A, Eddy S, Bateman A: Rfam: annotating non-coding RNAs in complete genomes. Nucleic Acids Res 2005, 33(Database issue):121-124.</p></li>
            <li><p>[NDB] Berman HM, Olson WK, Beveridge DL, Westbrook J, Gelbin A, Demeny T, Hsieh SH, Srinivasan AR, Schneider B: The nucleic acid database. A comprehensive relational database of three-dimensional structures of nucleic acids. Biophys J 1992, 63(3):751-759.</p></li>
            <li><p>[SILVA] Source: arb-silva, <a href="https://www.arb-silva.de/">https://www.arb-silva.de/</a>  used under a Creative Commons Attribution 4.0 International Public License (<a href="https://creativecommons.org/licenses/by/4.0/legalcode">https://creativecommons.org/licenses/by/4.0/legalcode</a>)</p></li>
            <li><p>Quast C, Pruesse E, Yilmaz P, Gerken J, Schweer T, Yarza P, Peplies J, Glockner FO (2013) The SILVA ribosomal RNA gene database project: improved data processing and web-based tools. Nucleic Acids Research 41:D590-D596</p></li>
            <li><p>Yilmaz P, Parfrey LW, Yarza P, Gerken J, Pruesse E, Quast C, Schweer T,Peplies J, Ludwig W, Glockner FO (2014) The SILVA and "All-species Living Tree Project (LTP)" taxonomic frameworks. Nucleic Acid Res. 42:D643-D648</p></li>
            <li><p>[LTP] Source: mmg-ltp, <a href="https://imedea.uib-csic.es/mmg/ltp/">https://imedea.uib-csic.es/mmg/ltp/</a></p></li>
            <li><p>[GTDB] Source: gtdb, <a href="https://gtdb.ecogenomic.org/">https://gtdb.ecogenomic.org/</a></p></li>
            <li><p>Parks, D.H., et al. (2021). "GTDB: an ongoing census of bacterial and archaeal diversity through a phylogenetically consistent, rank normalized and complete genome-based taxonomy." Nucleic Acids Research, <a href="https://doi.org/10.1093/nar/gkab776">https://doi.org/10.1093/nar/gkab776</a></p></li>
            <li><p>Rinke, C, et al. (2021). "A standardized archaeal taxonomy for the Genome Taxonomy Database Nature Microbiology 6, 946â€“959 (2021). <a href="https://doi.org/10.1038/s41564-021-00918-8">https://doi.org/10.1038/s41564-021-00918-8</a> </p></li>
            <li><p>Parks, D.H., et al. (2020). "A complete domain-to-species taxonomy for Bacteria and Archaea." Nature Biotechnology, <a href="https://doi.org/10.1038/s41587-020-0501-8">https://doi.org/10.1038/s41587-020-0501-8</a></p></li>
            <li><p>Parks, D.H., et al. (2018). "A standardized bacterial taxonomy based on genome phylogeny substantially revises the tree of life." Nature Biotechnology, 36: 996-1004.</p></li>
        </ul>
    <br>
    </div>
{% endblock %}


















